### ISSUE TITLE
Coq XML protocol unable to respond when handling compcert.cparser.Parser

### ISSUE THREAD

**Issue by jimyu94** [coq/coq#7034]
#### Version

Coq 8.7.1, CompCert 3.2

#### Operating system
macOS 10.13.3

#### Description of the problem

If you use 

```
coqtop -q -R lib compcert.lib -R common compcert.common -R x86_64 compcert.x86_64 -R x86 compcert.x86 -R backend compcert.backend -R cfrontend compcert.cfrontend -R driver compcert.driver -R flocq compcert.flocq -R exportclight compcert.exportclight -R cparser compcert.cparser -w -deprecated-implicit-arguments -w -deprecated-implicit-arguments -w -deprecated-implicit-arguments -ideslave -main-channel stdfds
```

and `Add` the sentences of `Parser.v` in the `cparser`folder, the whole `coqtop` process will freeze at a point, and `Ctrl+C` cannot end the process. This could result from the fact that `Parser.v` has more than 34,000 lines of code. However,

**Comment by ejgallego** [comment]
@jimyu94 can you run `coqtop` with the `-debug -stm-debug` options and send back the terminal output before the hang?

**Comment by jimyu94** [comment]
@ejgallego Sure! Do you mean to append `-debug -stm-debug` to the end of my Coq XML (ideslave) command, or do you mean to use `coqtop` instead of Coq XML?

**Comment by ejgallego** [comment]
Yeah I mean to call Coq in debug mode with `coqtop -debug -stm-debug -q -R lib...`.

That should give you lots of debug output. Then, we can see where the problem is.

**Comment by jimyu94** [comment]
[coqtop_plain.txt](https://github.com/coq/coq/files/1836256/coqtop_plain.txt)
[coqtop_debug.txt](https://github.com/coq/coq/files/1836257/coqtop_debug.txt)
I've pasted both the debug mode error and the plain `coqtop` error. Because I was using a Python script to execute it, I was not aware of the error output. But indeed 

```
coqc -q  -R lib compcert.lib  -R common compcert.common  -R x86_64 compcert.x86_64  -R x86 compcert.x86  -R backend compcert.backend  -R cfrontend compcert.cfrontend  -R driver compcert.driver  -R flocq compcert.flocq  -R exportclight compcert.exportclight  -R cparser compcert.cparser cparser/Parser.v
```

has no problem compiling it. Could you check why the error happened? Thanks! I've pasted the Parser.v as Parser_v.txt here.
[Parser_v.txt](https://github.com/coq/coq/files/1836272/Parser_v.txt)

**Comment by jimyu94** [comment]
@ejgallego

**Comment by gares** [comment]
Thanks for the debug info, it looks like a bug (parser gets out of sync), I'll try to reproduce it locally.

**Comment by gares** [comment]
I can't reproduce it here. I'm using CoqIDE and I jump to the end of Parser.v (I've checked out CompCert master, not the exact tag). Coq takes a few second to process almost every line in the file, and then starts doing the vm_compute stuff at the end of the file. No parse error.

I'm afraid you need tell me exactly what you mean by "Add the sentences", since I'm pretty sure you don't write XML by hand.

**Comment by maximedenes** [comment]
@gares this is a report for Coq 8.7.1. What version did you try to reproduce with?

**Comment by gares** [comment]
Oh, I've used Coq master. Let me try with 8.7

**Comment by gares** [comment]
The same on 8.7

**Comment by maximedenes** [comment]
> on 8.7

You mean 8.7.1?

**Comment by ejgallego** [comment]
Need to have a look, but I guess that CoqIDE doesn't trigger the exact same protocol calls than @jimyu94 , so it doesn't trigger the bug. I used to see many bugs like this in jsCoq, due to doing a long sequence of `Add`s.

Maybe the classifier doesn't force eager execution enough?

@jimyu94 , I guess the workaround for now is to call `Exec` after each `Add`. This is what `coqc` does and what jsCoq [sadly] is forced to do.

**Comment by ejgallego** [comment]
HA, `Program Instance` is involved. Color me surprised.

**Comment by gares** [comment]
BTW, the output files @jimyu94 has attached are quite odd. I see both `Coq >` lines, as if `coqtop` was used, but I also see `-ideslave` and I'm pretty sure the main loop of coqide does not print these lines.

I'm afraid we can't make progress without a way to reproduce it.

**Comment by ejgallego** [comment]
You can reproduce it using the strategy on https://github.com/ejgallego/coq-serapi/issues/47

Basically what `ReadFile` does is to call `Add` for the whole file.

**Comment by gares** [comment]
I did download coq-serapi (branch v8.8) and build it. Then I did `(ReadFile "cparser/Parser.v")` and all went good.  Then I tried to randomly ask `(Query ((sid 99)) Ast)` and I got an AST....

**Comment by gares** [comment]
I did not ask the AST for all senteces (they are 2600), is there a way do write a for loop in sertop?

**Comment by ejgallego** [comment]
> is there a way do write a for loop in sertop?

Nope, the next version of the query language does indeed allow for ranges, but that's not ready yet. You can try to `(Exec sid)` the last sentence maybe?

**Comment by ejgallego** [comment]
Of course you can always do a little of `echo "(ReadFile foo) `seq blah...`.." | sertop

**Comment by jimyu94** [comment]
Thanks for helping! I found that the files I provided above uses plain `coqtop`, which means I should escape `\r\n` but I forgot to do so. I found the freezing problem using `coqtop -ideslave`, which doesn't need to escape `\r\n`. I will provide the output of `ideslave` later today. May I ask you how to escape `\r\n` in the plain `coqtop` so I could also test the plain `coqtop`? @gares I also don't have problem with `(ReadFile "cparser/Parser.v")` in SerAPI. If I call 2600 Ast query in SerAPI, it freezes at a point of time. Also, if I issue 2600 `coqtop -ideslave` `Add`  calls, it will freeze (I will attach my inputs and the outputs later today!)

**Comment by ejgallego** [comment]
> in SerAPI. If I call 2600 Ast queries in SerAPI, it freezes at a point of time.

- Can you identify the exact sentence that creates the problem?
- Are you running `(Exec $last_sentence)` before calling `Query`.

**Comment by jimyu94** [comment]
@gares Here is my exact input to `coqtop -ideslave`. This time I used my Linux machine with Coq 8.7.1, and it has the same problem. Could you check it on your machine? Thanks!
[io.txt](https://github.com/coq/coq/files/1839850/io.txt)
[input.txt](https://github.com/coq/coq/files/1839851/input.txt)

@ejgallego No. I didn't use the Exec command before callng Query. I will try it right now!

**Comment by ejgallego** [comment]
Try 8.8 if you can too!

**Comment by jimyu94** [comment]
@ejgallego @gares Here is my input for SerAPI to trigger the freezing state in my Python code.
[input.txt](https://github.com/coq/coq/files/1840026/input.txt)
I am so sorry that after trying this input by just copying and pasting (instead of using my Python script), the freezing didn't happen, which means my Python code is to blame.

However, I can reproduce the freezing state in `coqtop -ideslave` with the input file that I shared with you an hour ago. To use it, you just need to first call

```
coqtop -q -R lib compcert.lib -R common compcert.common -R x86_64 compcert.x86_64 -R x86 compcert.x86 -R backend compcert.backend -R cfrontend compcert.cfrontend -R driver compcert.driver -R flocq compcert.flocq -R exportclight compcert.exportclight -R cparser compcert.cparser -w -deprecated-implicit-arguments -w -deprecated-implicit-arguments -w -deprecated-implicit-arguments -ideslave -main-channel stdfds
```

In the compcert directory, and then copy and paste all the `call`s in my input file.

**Comment by ejgallego** [comment]
@jimyu94 note, change your scripts to use `(Query ((sid $number)) Ast)` as this is the form 8.8 uses. Sorry for the non-compatible change.

Regarding the XML part, it is hard to say what's going on. I guess you can fully work in SerAPI for now until this issue is resolved.

**Comment by jimyu94** [comment]
@ejgallego Thanks for your suggestions!

**Comment by jimyu94** [comment]
@gares Hi! I've found the same problem in `mathcomp.algebra.vector`. Could you have a test? Here are my inputs and outputs. 
[io.txt](https://github.com/coq/coq/files/1865826/io.txt)
[input.txt](https://github.com/coq/coq/files/1865827/input.txt)
